{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/grabeda2/miniconda3/envs/similarity/lib/python3.10/site-packages/pandas/core/computation/expressions.py:21: UserWarning: Pandas requires version '2.8.0' or newer of 'numexpr' (version '2.7.3' currently installed).\n",
      "  from pandas.core.computation.check import NUMEXPR_INSTALLED\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import pickle\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "def load_object(filename):\n",
    "    with open(filename, 'rb') as f:\n",
    "        return pickle.load(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "19443"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "affinity_dict = load_object('/data/grbv/PDBbind/DTI5_general_affinity_dict.pkl')\n",
    "len(affinity_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "casf2016_dir = '/data/grbv/PDBbind/DTI_5/input_graphs_esm2_t6_8M/test_data/casf2016'\n",
    "casf2016_complexes = [filename[0:4] for filename in os.listdir(casf2016_dir) if 'graph' in filename]\n",
    "\n",
    "casf2013_dir = '/data/grbv/PDBbind/DTI_5/input_graphs_esm2_t6_8M/test_data/casf2013'\n",
    "casf2013_complexes = [filename[0:4] for filename in os.listdir(casf2013_dir) if 'graph' in filename]\n",
    "\n",
    "train_dir = '/data/grbv/PDBbind/DTI_5/input_graphs_esm2_t6_8M/training_data'\n",
    "train_complexes = [filename[0:4] for filename in os.listdir(train_dir) if 'graph' in filename]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Generate sequence dict with all complexes\n",
    "\n",
    "# sequence_dict = {}\n",
    "# save_path = '/data/grbv/PDBbind/DTI5_sequence_dict.pkl'\n",
    "\n",
    "\n",
    "# data_dir = '/data/grbv/PDBbind/DTI5_input_data_processed'\n",
    "\n",
    "# complexes = [folder for folder in os.listdir(data_dir) if len(folder)==4 and folder[0].isdigit()]\n",
    "# for compl in complexes: \n",
    "\n",
    "#     in_dataset = False\n",
    "#     if compl in train_complexes: in_dataset=True\n",
    "#     elif compl in casf2013_complexes: in_dataset=True\n",
    "#     elif compl in casf2016_complexes: in_dataset=True\n",
    "\n",
    "#     if in_dataset:\n",
    "#         path = os.path.join(data_dir, compl)\n",
    "\n",
    "#         for file in os.scandir(path):\n",
    "#             if file.name.endswith('protein_dict.pkl'):\n",
    "#                 id = file.name[0:4]       \n",
    "#                 protein_dict = load_object(file.path)\n",
    "\n",
    "#                 seq = ''\n",
    "#                 for chain in protein_dict:\n",
    "#                     s = protein_dict[chain]['aa_seq']\n",
    "#                     seq = seq + s\n",
    "\n",
    "#                 sequence_dict[id]=seq   \n",
    "\n",
    "# with open(save_path, 'wb') as fp:\n",
    "#     pickle.dump(sequence_dict, fp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "19132"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sequence_dict = load_object('/data/grbv/PDBbind/DTI5_sequence_dict.pkl')\n",
    "len(sequence_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Save sequences as fasta file:\n",
    "\n",
    "# fasta_file = '/data/grbv/PDBbind/DTI5_sequences.fasta'\n",
    "\n",
    "# with open(fasta_file, 'w') as f:\n",
    "#     for key in sequence_dict:\n",
    "#         f.write('>' + key + '\\n' + sequence_dict[key] + '\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "easy-cluster /data/grbv/PDBbind/DTI5_sequences.fasta clusterRes tmp --min-seq-id 0.8 -c 0.8 --cov-mode 1 \n",
      "\n",
      "MMseqs Version:                     \t15.6f452\n",
      "Substitution matrix                 \taa:blosum62.out,nucl:nucleotide.out\n",
      "Seed substitution matrix            \taa:VTML80.out,nucl:nucleotide.out\n",
      "Sensitivity                         \t4\n",
      "k-mer length                        \t0\n",
      "Target search mode                  \t0\n",
      "k-score                             \tseq:2147483647,prof:2147483647\n",
      "Alphabet size                       \taa:21,nucl:5\n",
      "Max sequence length                 \t65535\n",
      "Max results per query               \t20\n",
      "Split database                      \t0\n",
      "Split mode                          \t2\n",
      "Split memory limit                  \t0\n",
      "Coverage threshold                  \t0.8\n",
      "Coverage mode                       \t1\n",
      "Compositional bias                  \t1\n",
      "Compositional bias                  \t1\n",
      "Diagonal scoring                    \ttrue\n",
      "Exact k-mer matching                \t0\n",
      "Mask residues                       \t1\n",
      "Mask residues probability           \t0.9\n",
      "Mask lower case residues            \t0\n",
      "Minimum diagonal score              \t15\n",
      "Selected taxa                       \t\n",
      "Include identical seq. id.          \tfalse\n",
      "Spaced k-mers                       \t1\n",
      "Preload mode                        \t0\n",
      "Pseudo count a                      \tsubstitution:1.100,context:1.400\n",
      "Pseudo count b                      \tsubstitution:4.100,context:5.800\n",
      "Spaced k-mer pattern                \t\n",
      "Local temporary path                \t\n",
      "Threads                             \t40\n",
      "Compressed                          \t0\n",
      "Verbosity                           \t3\n",
      "Add backtrace                       \tfalse\n",
      "Alignment mode                      \t3\n",
      "Alignment mode                      \t0\n",
      "Allow wrapped scoring               \tfalse\n",
      "E-value threshold                   \t0.001\n",
      "Seq. id. threshold                  \t0.8\n",
      "Min alignment length                \t0\n",
      "Seq. id. mode                       \t0\n",
      "Alternative alignments              \t0\n",
      "Max reject                          \t2147483647\n",
      "Max accept                          \t2147483647\n",
      "Score bias                          \t0\n",
      "Realign hits                        \tfalse\n",
      "Realign score bias                  \t-0.2\n",
      "Realign max seqs                    \t2147483647\n",
      "Correlation score weight            \t0\n",
      "Gap open cost                       \taa:11,nucl:5\n",
      "Gap extension cost                  \taa:1,nucl:2\n",
      "Zdrop                               \t40\n",
      "Rescore mode                        \t0\n",
      "Remove hits by seq. id. and coverage\tfalse\n",
      "Sort results                        \t0\n",
      "Cluster mode                        \t0\n",
      "Max connected component depth       \t1000\n",
      "Similarity type                     \t2\n",
      "Weight file name                    \t\n",
      "Cluster Weight threshold            \t0.9\n",
      "Single step clustering              \tfalse\n",
      "Cascaded clustering steps           \t3\n",
      "Cluster reassign                    \tfalse\n",
      "Remove temporary files              \ttrue\n",
      "Force restart with latest tmp       \tfalse\n",
      "MPI runner                          \t\n",
      "k-mers per sequence                 \t21\n",
      "Scale k-mers per sequence           \taa:0.000,nucl:0.200\n",
      "Adjust k-mer length                 \tfalse\n",
      "Shift hash                          \t67\n",
      "Include only extendable             \tfalse\n",
      "Skip repeating k-mers               \tfalse\n",
      "Database type                       \t0\n",
      "Shuffle input database              \ttrue\n",
      "Createdb mode                       \t1\n",
      "Write lookup file                   \t0\n",
      "Offset of numeric ids               \t0\n",
      "\n",
      "createdb /data/grbv/PDBbind/DTI5_sequences.fasta tmp/1793533943949204301/input --dbtype 0 --shuffle 1 --createdb-mode 1 --write-lookup 0 --id-offset 0 --compressed 0 -v 3 \n",
      "\n",
      "Shuffle database cannot be combined with --createdb-mode 0\n",
      "We recompute with --shuffle 0\n",
      "Converting sequences\n",
      "[=\n",
      "Time for merging to input_h: 0h 0m 0s 0ms\n",
      "Time for merging to input: 0h 0m 0s 0ms\n",
      "Database type: Aminoacid\n",
      "Time for processing: 0h 0m 0s 17ms\n",
      "Create directory tmp/1793533943949204301/clu_tmp\n",
      "cluster tmp/1793533943949204301/input tmp/1793533943949204301/clu tmp/1793533943949204301/clu_tmp --max-seqs 20 -c 0.8 --cov-mode 1 --spaced-kmer-mode 1 --alignment-mode 3 -e 0.001 --min-seq-id 0.8 --remove-tmp-files 1 \n",
      "\n",
      "Set cluster sensitivity to -s 1.000000\n",
      "Set cluster mode GREEDY MEM\n",
      "Set cluster iterations to 1\n",
      "linclust tmp/1793533943949204301/input tmp/1793533943949204301/clu_tmp/846167229837907227/clu_redundancy tmp/1793533943949204301/clu_tmp/846167229837907227/linclust --cluster-mode 3 --max-iterations 1000 --similarity-type 2 --threads 40 --compressed 0 -v 3 --cluster-weight-threshold 0.9 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.8 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 1 --max-seq-len 65535 --comp-bias-corr 0 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --alph-size aa:13,nucl:5 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale aa:0.000,nucl:0.200 --adjust-kmer-len 0 --mask 0 --mask-prob 0.9 --mask-lower-case 0 -k 0 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --rescore-mode 0 --filter-hits 0 --sort-results 0 --remove-tmp-files 1 --force-reuse 0 \n",
      "\n",
      "kmermatcher tmp/1793533943949204301/input tmp/1793533943949204301/clu_tmp/846167229837907227/linclust/10827341071345083566/pref --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --alph-size aa:13,nucl:5 --min-seq-id 0.8 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale aa:0.000,nucl:0.200 --adjust-kmer-len 0 --mask 0 --mask-prob 0.9 --mask-lower-case 0 --cov-mode 1 -k 0 -c 0.8 --max-seq-len 65535 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --threads 40 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "kmermatcher tmp/1793533943949204301/input tmp/1793533943949204301/clu_tmp/846167229837907227/linclust/10827341071345083566/pref --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --alph-size aa:13,nucl:5 --min-seq-id 0.8 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale aa:0.000,nucl:0.200 --adjust-kmer-len 0 --mask 0 --mask-prob 0.9 --mask-lower-case 0 --cov-mode 1 -k 0 -c 0.8 --max-seq-len 65535 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --threads 40 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Database size: 19132 type: Aminoacid\n",
      "Reduced amino acid alphabet: (A S T) (C) (D B N) (E Q Z) (F Y) (G) (H) (I V) (K R) (L J M) (P) (W) (X) \n",
      "\n",
      "Generate k-mers list for 1 split\n",
      "[=================================================================] 19.13K 0s 65ms\n",
      "Sort kmer 0h 0m 0s 195ms\n",
      "Sort by rep. sequence 0h 0m 0s 168ms\n",
      "Time for fill: 0h 0m 0s 9ms\n",
      "Time for merging to pref: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 595ms\n",
      "rescorediagonal tmp/1793533943949204301/input tmp/1793533943949204301/input tmp/1793533943949204301/clu_tmp/846167229837907227/linclust/10827341071345083566/pref tmp/1793533943949204301/clu_tmp/846167229837907227/linclust/10827341071345083566/pref_rescore1 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --rescore-mode 0 --wrapped-scoring 0 --filter-hits 0 -e 0.001 -c 0.8 -a 0 --cov-mode 1 --min-seq-id 0.8 --min-aln-len 0 --seq-id-mode 0 --add-self-matches 0 --sort-results 0 --db-load-mode 0 --threads 40 --compressed 0 -v 3 \n",
      "\n",
      "[===========================================================] 19.13K 0s 47ms\n",
      "======Time for merging to pref_rescore1: 0h 0m 0s 9ms\n",
      "Time for processing: 0h 0m 0s 141ms\n",
      "clust tmp/1793533943949204301/input tmp/1793533943949204301/clu_tmp/846167229837907227/linclust/10827341071345083566/pref_rescore1 tmp/1793533943949204301/clu_tmp/846167229837907227/linclust/10827341071345083566/pre_clust --cluster-mode 3 --max-iterations 1000 --similarity-type 2 --threads 40 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Clustering mode: Greedy Low Mem\n",
      "Total time: 0h 0m 0s 127ms\n",
      "\n",
      "Size of the sequence database: 19132\n",
      "Size of the alignment database: 19132\n",
      "Number of clusters: 8270\n",
      "\n",
      "Writing results 0h 0m 0s 2ms\n",
      "Time for merging to pre_clust: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 138ms\n",
      "createsubdb tmp/1793533943949204301/clu_tmp/846167229837907227/linclust/10827341071345083566/order_redundancy tmp/1793533943949204301/input tmp/1793533943949204301/clu_tmp/846167229837907227/linclust/10827341071345083566/input_step_redundancy -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step_redundancy: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 3ms\n",
      "createsubdb tmp/1793533943949204301/clu_tmp/846167229837907227/linclust/10827341071345083566/order_redundancy tmp/1793533943949204301/clu_tmp/846167229837907227/linclust/10827341071345083566/pref tmp/1793533943949204301/clu_tmp/846167229837907227/linclust/10827341071345083566/pref_filter1 -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to pref_filter1: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 4ms\n",
      "filterdb tmp/1793533943949204301/clu_tmp/846167229837907227/linclust/10827341071345083566/pref_filter1 tmp/1793533943949204301/clu_tmp/846167229837907227/linclust/10827341071345083566/pref_filter2 --filter-file tmp/1793533943949204301/clu_tmp/846167229837907227/linclust/10827341071345083566/order_redundancy --threads 40 --compressed 0 -v 3 \n",
      "\n",
      "Filtering using file(s)\n",
      "[=================================================================] 8.27K 0s 58ms\n",
      "Time for merging to pref_filter2: 0h 0m 0s 5ms\n",
      "Time for processing: 0h 0m 0s 109ms\n",
      "rescorediagonal tmp/1793533943949204301/clu_tmp/846167229837907227/linclust/10827341071345083566/input_step_redundancy tmp/1793533943949204301/clu_tmp/846167229837907227/linclust/10827341071345083566/input_step_redundancy tmp/1793533943949204301/clu_tmp/846167229837907227/linclust/10827341071345083566/pref_filter2 tmp/1793533943949204301/clu_tmp/846167229837907227/linclust/10827341071345083566/pref_rescore2 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --rescore-mode 1 --wrapped-scoring 0 --filter-hits 1 -e 0.001 -c 0.8 -a 0 --cov-mode 1 --min-seq-id 0.8 --min-aln-len 0 --seq-id-mode 0 --add-self-matches 0 --sort-results 0 --db-load-mode 0 --threads 40 --compressed 0 -v 3 \n",
      "\n",
      "[======================================================] 8.27K 0s 30ms\n",
      "===========Time for merging to pref_rescore2: 0h 0m 0s 18ms\n",
      "Time for processing: 0h 0m 0s 121ms\n",
      "align tmp/1793533943949204301/clu_tmp/846167229837907227/linclust/10827341071345083566/input_step_redundancy tmp/1793533943949204301/clu_tmp/846167229837907227/linclust/10827341071345083566/input_step_redundancy tmp/1793533943949204301/clu_tmp/846167229837907227/linclust/10827341071345083566/pref_rescore2 tmp/1793533943949204301/clu_tmp/846167229837907227/linclust/10827341071345083566/aln --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.8 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 1 --max-seq-len 65535 --comp-bias-corr 0 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --threads 40 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 8270 type: Aminoacid\n",
      "Target database size: 8270 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 8.27K 1s 152ms\n",
      "Time for merging to aln: 0h 0m 0s 12ms\n",
      "18032 alignments calculated\n",
      "16095 sequence pairs passed the thresholds (0.892580 of overall calculated)\n",
      "1.946191 hits per query sequence\n",
      "Time for processing: 0h 0m 1s 210ms\n",
      "clust tmp/1793533943949204301/clu_tmp/846167229837907227/linclust/10827341071345083566/input_step_redundancy tmp/1793533943949204301/clu_tmp/846167229837907227/linclust/10827341071345083566/aln tmp/1793533943949204301/clu_tmp/846167229837907227/linclust/10827341071345083566/clust --cluster-mode 3 --max-iterations 1000 --similarity-type 2 --threads 40 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Clustering mode: Greedy Low Mem\n",
      "Total time: 0h 0m 0s 101ms\n",
      "\n",
      "Size of the sequence database: 8270\n",
      "Size of the alignment database: 8270\n",
      "Number of clusters: 4260\n",
      "\n",
      "Writing results 0h 0m 0s 1ms\n",
      "Time for merging to clust: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 108ms\n",
      "mergeclusters tmp/1793533943949204301/input tmp/1793533943949204301/clu_tmp/846167229837907227/clu_redundancy tmp/1793533943949204301/clu_tmp/846167229837907227/linclust/10827341071345083566/pre_clust tmp/1793533943949204301/clu_tmp/846167229837907227/linclust/10827341071345083566/clust --threads 40 --compressed 0 -v 3 \n",
      "\n",
      "Clustering step 1\n",
      "[=================================================================] 8.27K 0s 48ms\n",
      "Clustering step 2\n",
      "[=================================================================] 4.26K 0s 125ms\n",
      "Write merged clustering\n",
      "[=================================================================] 19.13K 0s 189ms\n",
      "Time for merging to clu_redundancy: 0h 0m 0s 3ms\n",
      "Time for processing: 0h 0m 0s 239ms\n",
      "rmdb tmp/1793533943949204301/clu_tmp/846167229837907227/linclust/10827341071345083566/pref_filter1 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb tmp/1793533943949204301/clu_tmp/846167229837907227/linclust/10827341071345083566/pref -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb tmp/1793533943949204301/clu_tmp/846167229837907227/linclust/10827341071345083566/pref_rescore1 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 1ms\n",
      "rmdb tmp/1793533943949204301/clu_tmp/846167229837907227/linclust/10827341071345083566/pre_clust -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb tmp/1793533943949204301/clu_tmp/846167229837907227/linclust/10827341071345083566/input_step_redundancy -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb tmp/1793533943949204301/clu_tmp/846167229837907227/linclust/10827341071345083566/input_step_redundancy_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb tmp/1793533943949204301/clu_tmp/846167229837907227/linclust/10827341071345083566/pref_filter2 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 1ms\n",
      "rmdb tmp/1793533943949204301/clu_tmp/846167229837907227/linclust/10827341071345083566/pref_rescore2 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 2ms\n",
      "rmdb tmp/1793533943949204301/clu_tmp/846167229837907227/linclust/10827341071345083566/aln -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 2ms\n",
      "rmdb tmp/1793533943949204301/clu_tmp/846167229837907227/linclust/10827341071345083566/clust -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "createsubdb tmp/1793533943949204301/clu_tmp/846167229837907227/clu_redundancy tmp/1793533943949204301/input tmp/1793533943949204301/clu_tmp/846167229837907227/input_step_redundancy -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step_redundancy: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 7ms\n",
      "prefilter tmp/1793533943949204301/clu_tmp/846167229837907227/input_step_redundancy tmp/1793533943949204301/clu_tmp/846167229837907227/input_step_redundancy tmp/1793533943949204301/clu_tmp/846167229837907227/pref_step0 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --seed-sub-mat 'aa:VTML80.out,nucl:nucleotide.out' -s 1 -k 0 --target-search-mode 0 --k-score seq:2147483647,prof:2147483647 --alph-size aa:21,nucl:5 --max-seq-len 65535 --max-seqs 20 --split 0 --split-mode 2 --split-memory-limit 0 -c 0.8 --cov-mode 1 --comp-bias-corr 0 --comp-bias-corr-scale 1 --diag-score 0 --exact-kmer-matching 0 --mask 1 --mask-prob 0.9 --mask-lower-case 0 --min-ungapped-score 0 --add-self-matches 0 --spaced-kmer-mode 1 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --threads 40 --compressed 0 -v 3 \n",
      "\n",
      "Query database size: 4260 type: Aminoacid\n",
      "Estimated memory consumption: 1003M\n",
      "Target database size: 4260 type: Aminoacid\n",
      "Index table k-mer threshold: 154 at k-mer size 6 \n",
      "Index table: counting k-mers\n",
      "[=================================================================] 4.26K 0s 162ms\n",
      "Index table: Masked residues: 11745\n",
      "Index table: fill\n",
      "[=================================================================] 4.26K 0s 46ms\n",
      "Index statistics\n",
      "Entries:          698446\n",
      "DB size:          492 MB\n",
      "Avg k-mer size:   0.010913\n",
      "Top 10 k-mers\n",
      "    LGLKFP\t120\n",
      "    VHFAQS\t114\n",
      "    WNGLGV\t109\n",
      "    SVCLFY\t99\n",
      "    DYPPVS\t60\n",
      "    KGSFAP\t59\n",
      "    ASKPFP\t59\n",
      "    TYCVKP\t59\n",
      "    TCVHSS\t59\n",
      "    YFEVSW\t59\n",
      "Time for index table init: 0h 0m 1s 61ms\n",
      "Process prefiltering step 1 of 1\n",
      "\n",
      "k-mer similarity threshold: 154\n",
      "Starting prefiltering scores calculation (step 1 of 1)\n",
      "Query db start 1 to 4260\n",
      "Target db start 1 to 4260\n",
      "[=================================================================] 4.26K 0s 87ms\n",
      "\n",
      "2.274136 k-mers per position\n",
      "1123 DB matches per sequence\n",
      "0 overflows\n",
      "5 sequences passed prefiltering per query sequence\n",
      "2 median result list length\n",
      "0 sequences with 0 size result lists\n",
      "Time for merging to pref_step0: 0h 0m 0s 4ms\n",
      "Time for processing: 0h 0m 1s 617ms\n",
      "align tmp/1793533943949204301/clu_tmp/846167229837907227/input_step_redundancy tmp/1793533943949204301/clu_tmp/846167229837907227/input_step_redundancy tmp/1793533943949204301/clu_tmp/846167229837907227/pref_step0 tmp/1793533943949204301/clu_tmp/846167229837907227/aln_step0 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.8 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 1 --max-seq-len 65535 --comp-bias-corr 0 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --threads 40 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 4260 type: Aminoacid\n",
      "Target database size: 4260 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 4.26K 1s 608ms\n",
      "Time for merging to aln_step0: 0h 0m 0s 4ms\n",
      "19082 alignments calculated\n",
      "9015 sequence pairs passed the thresholds (0.472435 of overall calculated)\n",
      "2.116197 hits per query sequence\n",
      "Time for processing: 0h 0m 1s 673ms\n",
      "clust tmp/1793533943949204301/clu_tmp/846167229837907227/input_step_redundancy tmp/1793533943949204301/clu_tmp/846167229837907227/aln_step0 tmp/1793533943949204301/clu_tmp/846167229837907227/clu_step0 --cluster-mode 3 --max-iterations 1000 --similarity-type 2 --threads 40 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Clustering mode: Greedy Low Mem\n",
      "Total time: 0h 0m 0s 104ms\n",
      "\n",
      "Size of the sequence database: 4260\n",
      "Size of the alignment database: 4260\n",
      "Number of clusters: 3464\n",
      "\n",
      "Writing results 0h 0m 0s 0ms\n",
      "Time for merging to clu_step0: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 111ms\n",
      "mergeclusters tmp/1793533943949204301/input tmp/1793533943949204301/clu tmp/1793533943949204301/clu_tmp/846167229837907227/clu_redundancy tmp/1793533943949204301/clu_tmp/846167229837907227/clu_step0 --threads 40 --compressed 0 -v 3 \n",
      "\n",
      "Clustering step 1\n",
      "[=================================================================] 4.26K 0s 45ms\n",
      "Clustering step 2\n",
      "[=================================================================] 3.46K 0s 112ms\n",
      "Write merged clustering\n",
      "[=================================================================] 19.13K 0s 165ms\n",
      "Time for merging to clu: 0h 0m 0s 10ms\n",
      "Time for processing: 0h 0m 0s 202ms\n",
      "rmdb tmp/1793533943949204301/clu_tmp/846167229837907227/clu_redundancy -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb tmp/1793533943949204301/clu_tmp/846167229837907227/input_step_redundancy -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb tmp/1793533943949204301/clu_tmp/846167229837907227/input_step_redundancy_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb tmp/1793533943949204301/clu_tmp/846167229837907227/pref_step0 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 1ms\n",
      "rmdb tmp/1793533943949204301/clu_tmp/846167229837907227/aln_step0 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 2ms\n",
      "rmdb tmp/1793533943949204301/clu_tmp/846167229837907227/clu_step0 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "createtsv tmp/1793533943949204301/input tmp/1793533943949204301/input tmp/1793533943949204301/clu tmp/1793533943949204301/cluster.tsv --threads 40 -v 3 \n",
      "\n",
      "Time for merging to cluster.tsv: 0h 0m 0s 17ms\n",
      "Time for processing: 0h 0m 0s 86ms\n",
      "result2repseq tmp/1793533943949204301/input tmp/1793533943949204301/clu tmp/1793533943949204301/clu_rep --db-load-mode 0 --compressed 0 --threads 40 -v 3 \n",
      "\n",
      "[=================================================================] 3.46K 0s 37ms\n",
      "Time for merging to clu_rep: 0h 0m 0s 16ms\n",
      "Time for processing: 0h 0m 0s 80ms\n",
      "result2flat tmp/1793533943949204301/input tmp/1793533943949204301/input tmp/1793533943949204301/clu_rep tmp/1793533943949204301/rep_seq.fasta --use-fasta-header -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 16ms\n",
      "createseqfiledb tmp/1793533943949204301/input tmp/1793533943949204301/clu tmp/1793533943949204301/clu_seqs --threads 40 -v 3 \n",
      "\n",
      "[=================================================================] 3.46K 0s 64ms\n",
      "Time for merging to clu_seqs: 0h 0m 0s 4ms\n",
      "Time for processing: 0h 0m 0s 113ms\n",
      "result2flat tmp/1793533943949204301/input tmp/1793533943949204301/input tmp/1793533943949204301/clu_seqs tmp/1793533943949204301/all_seqs.fasta -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 52ms\n",
      "rmdb tmp/1793533943949204301/input -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb tmp/1793533943949204301/input_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb tmp/1793533943949204301/clu_seqs -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 6ms\n",
      "rmdb tmp/1793533943949204301/clu_rep -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 1ms\n",
      "rmdb tmp/1793533943949204301/clu -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 1ms\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "CompletedProcess(args=['mmseqs', 'easy-cluster', '/data/grbv/PDBbind/DTI5_sequences.fasta', 'clusterRes', 'tmp', '--min-seq-id', '0.8', '-c', '0.8', '--cov-mode', '1'], returncode=0)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#using mmseqs2 to cluster the sequences\n",
    "import subprocess\n",
    "\n",
    "#mmseqs easy-cluster test_data.fasta clusterRes tmp --min-seq-id 0.5 -c 0.8 --cov-mode 1\n",
    "\n",
    "subprocess.run(['mmseqs', 'easy-cluster', '/data/grbv/PDBbind/DTI5_sequences.fasta', 'clusterRes', 'tmp', '--min-seq-id', '0.8', '-c', '0.8', '--cov-mode', '1'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1bcj</td>\n",
       "      <td>1bcj</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2brp</td>\n",
       "      <td>2brp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2brp</td>\n",
       "      <td>1f9g</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>5gmh</td>\n",
       "      <td>5gmh</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3tww</td>\n",
       "      <td>3tww</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      0     1\n",
       "0  1bcj  1bcj\n",
       "1  2brp  2brp\n",
       "2  2brp  1f9g\n",
       "3  5gmh  5gmh\n",
       "4  3tww  3tww"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#lets analyse the clusterRes file\n",
    "clusterRes = pd.read_csv('clusterRes_cluster.tsv', sep='\\t', header=None)\n",
    "clusterRes.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3464 number of clusters\n"
     ]
    }
   ],
   "source": [
    "'''All members of the clustering are listed line by line. \n",
    "Each cluster is a consecutive block in the file. The first column always contains the representative sequence, the second contains the cluster member. For the example the cluster with the representative sequence Q0KJ32 contains four members itself and C0W539, D6KVP9, D1Y890.\n",
    " IDs are parsed from the header from the input database (see id parsing from headers).\n",
    "'''\n",
    "#lets count the number of clusters and its memebers\n",
    "print(len(clusterRes[0].unique()), 'number of clusters')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0\n",
       "5kr2    438\n",
       "5doh    420\n",
       "3kmx    342\n",
       "2v22    223\n",
       "2tpi    194\n",
       "       ... \n",
       "1mfa      1\n",
       "6bvb      1\n",
       "4np9      1\n",
       "4ris      1\n",
       "5xhz      1\n",
       "Name: count, Length: 3464, dtype: int64"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#lets count the number of sequences in each cluster\n",
    "clusterRes[0].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "import csv\n",
    "\n",
    "with open('clusterRes_cluster.tsv', 'r', newline='') as infile, open('clusterRes_cluster_DTI5_1.tsv', 'w', newline='') as outfile:\n",
    "    reader = csv.reader(infile, delimiter='\\t')\n",
    "    writer = csv.writer(outfile, delimiter='\\t')\n",
    "\n",
    "    for row in reader:\n",
    "\n",
    "        # Retrieve the value from the dictionary based on the second column\n",
    "        id = row[1]\n",
    "\n",
    "        if id in train_complexes: string = 'training'\n",
    "        if id in casf2016_complexes: string = 'test    '\n",
    "        if id in casf2013_complexes: string = 'test    '\n",
    "\n",
    "        affinity = affinity_dict[id]['log_kd_ki']\n",
    "\n",
    "        # Add this value as a third column\n",
    "        row.append(string)\n",
    "        row.append(affinity)\n",
    "\n",
    "        \n",
    "        # Write the modified row to the output file\n",
    "        writer.writerow(row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "import networkx as nx\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# Create a graph\n",
    "G = nx.Graph()\n",
    "\n",
    "# Add edges from the DataFrame\n",
    "for index, row in clusterRes.iterrows():\n",
    "    G.add_edge(row[0], row[1])\n",
    "\n",
    "# Set node positions using the spring layout\n",
    "pos = nx.spring_layout(G)\n",
    "# Set a clear background\n",
    "plt.figure(figsize=(50, 50))\n",
    "plt.gca().set_facecolor('white')\n",
    "\n",
    "# Draw nodes with adjusted size and color\n",
    "nx.draw_networkx_nodes(G, pos, node_size=200, node_color='skyblue', alpha=0.8)\n",
    "\n",
    "# Draw edges with adjusted thickness and color\n",
    "nx.draw_networkx_edges(G, pos, width=1.0, alpha=0.5, edge_color='gray')\n",
    "\n",
    "# Draw labels with adjusted font size\n",
    "nx.draw_networkx_labels(G, pos, font_size=8, font_color='black')\n",
    "\n",
    "\n",
    "# Show the plot\n",
    "plt.axis('off')  # Turn off axis\n",
    "plt.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "welo5",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
